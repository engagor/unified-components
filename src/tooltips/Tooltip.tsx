import React, { isValidElement, PropsWithChildren, ReactElement, useState } from 'react';
import Tippy from "@tippyjs/react/headless";
import Placement from "./Placement";
import Arrow from "./Arrow";

export type TooltipProps = PropsWithChildren<
    {
        content: string | ReactElement;
        placement?: Placement;
        visible?: boolean;
    }
>;

export default function Tooltip({ content, placement, visible, children }: TooltipProps): ReactElement | null {
    const [arrow, setArrow] = useState<HTMLDivElement | null>(null);

    if (!placement) {
        placement = 'top';
    }

    if (!isValidElement(children)) {
        return null;
    }

    const popperOptions = {
        modifiers: [
            { name: 'arrow', options: { element: arrow } },
        ],
    };

    const className = [
        'position-relative'
    ];

    return (
        <Tippy
            placement={placement}
            visible={visible}
            popperOptions={popperOptions}
            className={className.join(' ')}
            render={
                (attrs) => {
                    return (
                        <article className="bg-charcoal-color-400 box-shadow-default rounded-4 w-max-280 p-8" role="tooltip" {...attrs}>
                            <section className="text-base font-normal text-white">{content}</section>
                            <Arrow ref={(node) => setArrow(node)} className="text-charcoal-color-400" />
                        </article>
                    );
                }
            }
        >
            {children}
        </Tippy>
    );
}