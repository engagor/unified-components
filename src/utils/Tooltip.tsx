import { cloneElement, FunctionComponent, ReactElement, useEffect, useRef } from 'react';
import { Options } from 'tooltip.js';

const Tooltip: FunctionComponent<TooltipProps> = (props: TooltipProps): ReactElement => {
    const tooltipInstance: any = useRef(null);
    const tooltipRef: any = useRef(null);

    useEffect(
        (): () => void => {
            if (props.title && tooltipRef.current) {
                import(/* webpackChunkName: "tooltip-js" */ 'tooltip.js').then(
                    ({ default: TooltipLib }): void => {
                        tooltipInstance.current = new TooltipLib(
                            tooltipRef.current,
                            {
                                placement: props.placement,
                                title: props.title,
                                trigger: props.trigger,
                                html: props.html,
                                container: props.container,
                                boundariesElement: 'window',
                                template: props.template || '<div class="popper-tooltip" role="tooltip"><div class="tooltip-arrow"></div><div class="tooltip-inner"></div></div>'
                            }
                        );
                    }
                );
            }

            return (): void => {
                if (tooltipInstance.current !== null) {
                    tooltipInstance.current.dispose();
                }
            };
        },
        [
            props.placement,
            props.title,
            props.trigger,
            props.html,
            props.container,
            props.template,
            tooltipRef
        ]
    );

    // @ts-ignore
    const trigger: Trigger = props.children;

    if (!props.title) {
        return trigger;
    }

    return cloneElement(
        trigger,
        {
            ...trigger.props,
            ref: (node: any): void => {
                tooltipRef.current = node;

                if (typeof trigger.ref === 'function') {
                    trigger.ref(node);
                }
            }
        }
    );
};

interface Trigger extends ReactElement {
    ref: (node: ReactElement) => {};
}

interface TooltipProps {
    title: Options['title'];
    placement?: Options['placement'];
    trigger?: Options['trigger'];
    html?: Options['html'];
    container?: Options['container'];
    children: ReactElement;
    template?: string;
}

Tooltip.defaultProps = {
    placement: 'top',
    trigger: 'hover focus',
    html: false,
    container: document.body,
};

export default Tooltip;
