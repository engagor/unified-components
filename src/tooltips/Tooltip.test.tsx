import { renderToStaticMarkup } from 'react-dom/server';

jest.mock(
    'react-dom',
    () => {
        return {
            createPortal: (element: ReactElement, node: HTMLElement) => {
                node.innerHTML = renderToStaticMarkup(element);
            },
        }
    }
);

import React, { ReactElement } from 'react';
import renderer from 'react-test-renderer';
import Tooltip from './Tooltip';

it('renders correctly', async () => {
    const element = document.createElement('div');

    const tooltip = await renderer
        .create(
            (
                <Tooltip content="Hello this is my awesome tooltip!" visible={true} appendTo={element}>
                    <button>Trigger</button>
                </Tooltip>
            )
        )
        .toJSON();

    await expect(tooltip).toMatchSnapshot();
    expect(element).toMatchSnapshot();
});
