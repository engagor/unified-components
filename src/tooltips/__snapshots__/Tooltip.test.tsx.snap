// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<button>
  Trigger
</button>
`;

exports[`renders correctly 2`] = `
<div>
  <div
    data-tippy-root=""
    id="tippy-1"
    style="pointer-events: none; z-index: 9999; transition: none; position: absolute; left: 0px; top: 0px; margin: 0px; bottom: 0px; transform: translate(0px, -10px);"
  >
    <article
      class="bg-charcoal-color-400 shadow rounded-4 w-max-280 p-8"
      data-escaped=""
      data-placement="top"
      data-reference-hidden=""
      role="tooltip"
    >
      <section
        class="text-base font-normal text-white"
      >
        Hello this is my awesome tooltip!
      </section>
      <div
        class="arrow text-charcoal-color-400"
      />
    </article>
  </div>
</div>
`;
